cloudevents/work/agent/client/manifestwork.go

```diff
@@ -0,0 +1,195 @@
+	}
+
+	if resourceVersion <= lastResourceVersion {
+		klog.Infof("The work %s resource version is less than or equal to cached, ignore", manifestWork.Name)
+		return manifestWork, nil
+	}
+
+	updatedWork := manifestWork.DeepCopy()
+
+	// restore the fields that are maintained by local agent
+	updatedWork.Labels = lastWork.Labels
+	updatedWork.Annotations = lastWork.Annotations
+	updatedWork.Finalizers = lastWork.Finalizers
+	updatedWork.Status = lastWork.Status
+
+	c.watcher.Receive(watch.Event{
+		Type:   watch.Modified,
+		Object: updatedWork,
+	})
+
+	return updatedWork, nil
+}
+
+func (c *ManifestWorksAgentClient) UpdateStatus(ctx context.Context, manifestWork *workv1.ManifestWork, opts metav1.UpdateOptions) (*workv1.ManifestWork, error) {
+	klog.V(4).Infof("updating manifestwork %s status", manifestWork.Name)
+
+	_, err := c.lister.Get(manifestWork.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	eventDataType, err := types.ParseCloudEventsDataType(manifestWork.Annotations[codec.CloudEventsDataTypeAnnotationKey])
+	if err != nil {
+		return nil, err
+	}
+
+	updatedWork := manifestWork.DeepCopy()
+
+	eventType := types.CloudEventsType{
+		CloudEventsDataType: *eventDataType,
+		SubResource:         types.SubResourceStatus,
+		Action:              UpdateRequestAction,
+	}
+
+	if err := c.cloudEventsClient.Publish(ctx, eventType, updatedWork); err != nil {
+		return nil, err
+	}
+	return updatedWork, nil
+}
+
+func (c *ManifestWorksAgentClient) Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error {
+	klog.V(4).Infof("deleting manifestwork %s", name)
+	work, err := c.lister.Get(name)
+	if err != nil {
+		return err
+	}
+
+	c.watcher.Receive(watch.Event{Type: watch.Modified, Object: work})
+	return nil
+}
+
+func (c *ManifestWorksAgentClient) DeleteCollection(ctx context.Context, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error {
+	klog.Fatal("DeleteCollection function for ManifestWorksAgentClient has not been implemented")
+	return nil
+}
+
+func (c *ManifestWorksAgentClient) Get(ctx context.Context, name string, opts metav1.GetOptions) (*workv1.ManifestWork, error) {
+	klog.V(4).Infof("getting manifestwork %s", name)
+	return c.lister.Get(name)
+}
+
+func (c *ManifestWorksAgentClient) List(ctx context.Context, opts metav1.ListOptions) (*workv1.ManifestWorkList, error) {
+	klog.V(4).Infof("sync manifestworks")
+	// send resync request to fetch manifestworks from source
+	if err := c.cloudEventsClient.Resync(ctx); err != nil {
+		return nil, err
+	}
+
+	return &workv1.ManifestWorkList{}, nil
+}
+
+func (mw *ManifestWorksAgentClient) Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
+	return mw.watcher, nil
+}
+
+func (mw *ManifestWorksAgentClient) Patch(ctx context.Context, name string, pt kubetypes.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *workv1.ManifestWork, err error) {
```

Comment:
we have starting to use patch a lot in ocm repo.

Reply:
add todo